vperfetto-sdk.cpp

```diff
@@ -767,22 +825,53 @@ static uint64_t deriveGuestTimeDiffWithGuestAbsoluteTime(
     uint64_t diff = guestBootTimeNs - hostStartTimeNs;
 
     fprintf(stderr, "%s: time diff: %llu\n", __func__, (unsigned long long)diff);
-    return 0;
+    return diff;
 }
 
 static uint64_t deriveGuestTimeDiff(
     const std::vector<char>& guestTrace,
-    const std::vector<char>& hostTrace) {
+    const std::vector<char>& hostTrace,
+    int64_t tscOffset) {
 
     fprintf(stderr, "%s: Deriving guest time diff from guest and host traces\n", __func__);
 
+    // First check for CPU time sync data in both traces.
+    TraceCpuTimeSync hostSync, guestSync;
+    bool hasHostSync = getTraceCpuTimeSync(hostTrace, &hostSync);
+    bool hasGuestSync = getTraceCpuTimeSync(guestTrace, &guestSync);
+    bool sameClock = hasHostSync && hasGuestSync && hostSync.clockId == guestSync.clockId;
+    if (hasHostSync && hasGuestSync && sameClock) {
+        // Transform guest cpuTime to host:
+        guestSync.cpuTime -= tscOffset;
+        fprintf(stderr, "%s: CPU cycles/nanos: host %f, guest %f\n", __func__, hostSync.cpuCyclesPerNano,
+                guestSync.cpuCyclesPerNano);
+        // Guest and host frequency measurement should match.
+        double cyclesPerNano = hostSync.cpuCyclesPerNano;
```

Comment:
Should we check and warn if the guest and host frequency measurement from getTraceCpuTimeSync is way off?

Reply:
Done.